### ISSUE TITLE
jwt tokens

### ISSUE THREAD

**Issue by basz** [TomHAnderson/zf-oauth2-doctrine#29]
I tried to work with jwt tokens (unsuccessfully still) one thing i noticed the tokens are much longer then 40 chars. columns should be text to be able to query them by token

if you agree I'l update the orm config and PR

**Comment by TomHAnderson** [comment]
Did you follow the unit test?  If it works there you should be able to get it working.  I know JWT is hard and I relied on the unit tests myself to know if they work.

**Comment by basz** [comment]
hehe, I hardly understand the oauth concepts at this point. Which one should use when and why i would want JWT (contain data and that can be asserted without an 'expensive' db call) mix that with some apigility and oauth server bugs regarding jwt's and I'm back at regular tokens (yeah i'm a pussy!). For now at least... so much more to learn... (ember js for one, doctrine/skipper workflow for two,...)

Do you agree the access_token table should be able to contain the full JWT tokens? (thus, from char to text).

**Comment by TomHAnderson** [comment]
No.  The jwt isn't an access token.  It's a very different way to authenticate.  That's why my ERD separates them physically.  Good luck!  Right now I don't think AccessToken needs a change.

**Comment by basz** [comment]
ok understood.

However .... Apigility (or Oauth server) does store the complete JWT 'header.payload.checksum' token in the accessToken table whenever you set zf-oauth2->options->use_jwt_access_tokens = true. 
Why is that then, or is it an artifact and not important as i suspect.

**Comment by TomHAnderson** [comment]
Well, a checksum would fit, yes?  I don't know the answer to whether it's an artifact or not.

**Comment by TomHAnderson** [comment]
basz, does #30 0.2.4 fix this?

**Comment by basz** [comment]
#30 fixes the JWTBearer logic regarding the jti identifier and (reply protection) (partly due to my earlier fix in renaming expiration to expire).

The issue discussed here is about when the server creates JWT tokens (the option use_jwt_access_tokens  instruct the server to do that). Those can be really long and cannot be stored completely within the char field of the AccessToken table.

I have turned that option to off for now and are beginning to see the how and what of JWT (and oauth in general). It's how I learn the ins and out, by going through the code. Every now and then I find something that doesn't work. The only way to figure out if it's me or some lib is ask... :-) So my thanks goes out to you for taking the time to respond en educate me in the process!

**Comment by TomHAnderson** [comment]
Can you produce a unit test which shows the JWT exceeding the access_token field length?  To create this module I based it on the SQL from zf-oauth2
```sql
CREATE TABLE oauth_access_tokens (
    access_token VARCHAR(40) NOT NULL,
    client_id VARCHAR(80) NOT NULL,
    user_id VARCHAR(255),
    expires TIMESTAMP NOT NULL,
    scope VARCHAR(2000),
    CONSTRAINT access_token_pk PRIMARY KEY (access_token)
);
``` 
but the provided adapters in zf-oauth2 don't implement JWT.  

Once you have a unit test showing it breaking I'll change the field to a text ASAP

**Comment by TomHAnderson** [comment]
The change to text is staged in #31

**Comment by TomHAnderson** [comment]
I was able to duplicate the problem when use_jwt_access_tokens = true is used of > 255 characters in AccessToken.accessToken.

To do this I created command line tools and service to generate a JWT in zf-oauth2-client.
I also created command line tools to create a JWT and PublicKey records in zf-oauth2-doctrine.
While doing this I found the code to get a PublicKey for a Client assumes a 1:1 relationship: https://github.com/TomHAnderson/zf-oauth2-doctrine/blob/master/src/Adapter/DoctrineAdapter.php#L1318

The schema currently is 1:many I'll change this to 1:1 while changing accessToken to text.  I don't see an advantage of storing the PublicKey data in the Client table so a 1:1 relationship is acceptable.
